### ISSUE TITLE
Border vertex is getting collapsed when edge collapsing is turned off

### ISSUE THREAD

**Issue by Zylann** [zeux/meshoptimizer#312]
I have the following mesh:

![image](https://user-images.githubusercontent.com/1311555/124399216-fdf7f380-dd11-11eb-93ba-deae47e717d5.png)

My version of the library includes the change I made in https://github.com/zeux/meshoptimizer/issues/311#issuecomment-873502795 which is supposed to prevent any border vertices from being collapsed.

Running `meshopt_simplify` with a target index count set to 0 and error set to 0.0025, a dent appears:
![image](https://user-images.githubusercontent.com/1311555/124399276-3992bd80-dd12-11eb-80eb-edcd67f6c368.png)

Wireframe comparison:
![image](https://user-images.githubusercontent.com/1311555/124399318-78c10e80-dd12-11eb-9e8a-88b1f67d6e55.png)

It sounds like this should not have happened.

For reproduction purposes:
This model was generated at runtime and I don't have code around to dump it as a standard format, so I dumped its vertex array as binary:
[dump.zip](https://github.com/zeux/meshoptimizer/files/6760573/dump.zip)
First 32-bit integer is vertex count, second 32-bit integer is index count, then follows positions array, then normals array, then index array.
Code I used to write it:
```gdscript
	f.store_32(len(positions))
	f.store_32(len(indices))
	
	for i in len(positions):
		f.store_float(positions[i].x)
		f.store_float(positions[i].y)
		f.store_float(positions[i].z)

	for i in len(normals):
		f.store_float(normals[i].x)
		f.store_float(normals[i].y)
		f.store_float(normals[i].z)

	for i in len(indices):
		f.store_32(indices[i])
```

**Comment by zeux** [comment]
It would be a bit easier to investigate this if the input was an obj file; it should be as easy to output as a binary file - just output “v” for each vertex position and “f” for each triangle using 1-based indices.

**Comment by Zylann** [comment]
Yeah I was wary of turning it into text because I thought it could break determinism, but I can give it a try

**Comment by Zylann** [comment]
Here is what I could get:
[dump.zip](https://github.com/zeux/meshoptimizer/files/6760687/dump.zip)

**Comment by zeux** [comment]
Thanks! I initially thought that classification is to blame here but it looks like the border is correctly classified as, well, border:

![image](https://user-images.githubusercontent.com/1106629/125009889-4832ea80-e01a-11eb-8c20-4e71714fd20b.png)

(blue=border)

Mesh before simplification:

![image](https://user-images.githubusercontent.com/1106629/125010146-d018f480-e01a-11eb-84c6-958a1921a535.png)

Mesh after simplification without any patches:

![image](https://user-images.githubusercontent.com/1106629/125010207-ec1c9600-e01a-11eb-851d-0b526358a214.png)

Mesh after simplification to 1% of triangles while disabling border collapses:

![image](https://user-images.githubusercontent.com/1106629/125010268-0eaeaf00-e01b-11eb-8ac0-cd39e7a0df3b.png)

It's a little difficult to see because of the different camera angles but it seems to me that this is all correct, and there's no dent. 

This is using gltfpack to simplify the .obj file. It's not quite the same as what you do because this welds vertices with the same position (since obj file doesn't have anything else), so it produces a different indexing. I will need to hack gltfpack to use the exact same indexing as present in the original file to analyze this more carefully...

**Comment by zeux** [comment]
Aha - ok - with the original index data and without any post-processing the classification does not recognize one of the vertices there as the border, hence the error in simplification.

![image](https://user-images.githubusercontent.com/1106629/125011381-34d54e80-e01d-11eb-8019-581e2ea1f870.png)

**Comment by zeux** [comment]
Alright phew - got it. This mesh had degenerate triangles in the input along the border in some places. These triangles really confused the classification code, as there are some shortcuts being taken right now that sort of assume this doesn't happen, which is of course a mistake. I need to figure out what the best way is to deal with these, but after quick patch I get this:

![image](https://user-images.githubusercontent.com/1106629/125013258-5edc4000-e020-11eb-85ef-454cbafef268.png)

The extra layer of quads on the border is due to the quick patch locking down vertices of degenerate triangles, which is probably too aggressive. I'll need to think about this a bit wrt what the correct solution is but I should be able to fix this later this week.

**Comment by Zylann** [comment]
I'll need to investigate if my implementation of Transvoxel can avoid producing degenerate triangles.

I assume the border of quads actually occurs because my implementation is not able to share vertices in one of the axis directions (because it starts a new deck and has no "previous" at this stage), which results in duplicated vertices along the border.

Simplification takes almost twice the time Transvoxel takes so at the moment I considered running a minimum of optimization steps.
